README.md
# pyDirectionCompSense

Compressed sensing for directional data. This dataset will be written from scratch, attempting to do so in Cython, so that we can have a code that is optimal for data that is directional in nature (e.g. diffusion data). By doing this, we have a chance at creating our own method for solving the compressed sensing minimization problem.

The idea is that we will write the code in a modular fashion, looking for where to optimize when we need to. As per Jason Lerch, "The first rule of optimization... Don't", implying that we should only optimize when it's absolutely necessary. 

The code will first be written in Python and then profiled -- all of this being noted in the commits (hopefully), in order to get things up and running. The plan is to:
  <b> Proof of Principle
  
    1) Have the code be able to properly perform TV operations on single slice (2D) datasets
    
    2) Have the code properly run FFT's (including knowledge of all normalization factors) as well as XFMs (Wavelet), and calculate the gradients for them
    
    3) Be able to feed this information in, along with pseudo-undersampled data of high SNR phantoms into an optimization algorithm (scipy.minimize to start) utilizing CS reconstruction techniques and compare to the gold standard (fully sampled) case
  
  <b> Secondary Testing 
  
    1) Build the foundation of the directional term in two ways:
    
      a. Using the method of a weighting term that is calculated via some f(dot(di,dj)) -- most likely a Gaussian term
      
      b. Using the method of an LSQ fitting technique with the images around it, using A as a matrix of the differences of the vector directions.
    
    2) Test the Diffusion phantom data without the directional terms
    
    3) Test the Diffusion phantom data with the directional terms and compare to each other, without the term, and the gold standard (fully sampled case)
  
  <b> Tertiary Testing
  
    1) Actually undersample data from the scanner
    
    2) Test reconstruction on this data -- both with and without the directional term (depending on the data that we're working with, if it's dMRI or just plain MRI)
    
    


history.txt
 2/1: a =
 2/2: a =CC
 4/1: a = "foo"
 4/2: import recon_CS
 4/3: recon_CS()
 4/4: recon_CS.recon_CS()
 4/5: reload(recon_CS)
 4/6: recon_CS.recon_CS()
 4/7: reload(recon_CS)
 4/8: recon_CS.recon_CS()
 4/9: a = recon_CS.recon_CS()
4/10: a
4/11: import matplotlib.pyplot as plt
4/12: a['x']
4/13: im = a['x']
4/14: im
4/15: im.shape = (256,256)
4/16: im
4/17: plt.imshow(abs(im))
4/18: plt.show()
4/19: recon_CS.recon_CS(TVWeight=0)
4/20: recon_CS.recon_CS(XFMWeight=0)
4/21: recon_CS.recon_CS()
4/22: reload(recon_CS)
4/23: recon_CS.recon_CS()
4/24: reload(recon_CS)
4/25: recon_CS.recon_CS()
4/26: reload(recon_CS)
4/27: recon_CS.recon_CS()
4/28: recon_CS.recon_CS(XFMWeight=1)
4/29: x = recon_CS.recon_CS(XFMWeight=1)
4/30: plt.imshow(abs(x['x']))
4/31: abs(x['x']).shape
4/32: abs(x['x']).shape = (256,256)
4/33: plt.imshow(abs(x['x']))
4/34: x['x'].shape = (256,256)
4/35: plt.imshow(abs(x['x']))
4/36: plt.show()
4/37: x = recon_CS.recon_CS(XFMWeight=0.1)
4/38: x = recon_CS.recon_CS(XFMWeight=0.5)
4/39: reload(recon_CS)
4/40: x = recon_CS.recon_CS(XFMWeight=0.5)
4/41: reload(recon_CS)
4/42: x = recon_CS.recon_CS(XFMWeight=0.5)
4/43: x = recon_CS.recon_CS(XFMWeight=0.5,TVWeight=0)
4/44: reload(recon_CS)
4/45: x = recon_CS.recon_CS(XFMWeight=0.5,TVWeight=0)
4/46: print('obj %.4d',1.23)
4/47: print('obj %4d',1.23)
4/48: print('obj %d',1.23)
4/49: print('obj %.2f',%1.23)
4/50: print('obj %.2f' %1.23)
4/51: reload(recon_CS)
4/52: x = recon_CS.recon_CS(XFMWeight=0.5,TVWeight=0)
 5/1: import rwt
 5/2: rwt.__file__
 5/3: rwt.__package__
 5/4: rwt.__path__
 5/5: x = recon_CS.recon_CS(XFMWeight=0.5,TVWeight=0)
 5/6: obj: 3140.25
 5/7: tv: 0.00
 5/8: xfm: 24461.07
 5/9: obj: 5536.49
5/10: tv: 0.00
5/11: xfm: 106697.66
5/12: obj: 3151.09
5/13: tv: 0.00
5/14: xfm: 27572.97
5/15: obj: 3140.56
5/16: tv: 0.00
5/17: xfm: 24585.67
5/18: obj: 3140.26
5/19: tv: 0.00
5/20: xfm: 24466.13
5/21: obj: 3140.25
5/22: tv: 0.00
5/23: xfm: 24461.27
5/24: obj: 3140.25
5/25: tv: 0.00
5/26: xfm: 24461.07
5/27: obj: 3140.25
5/28: tv: 0.00
5/29: xfm: 24461.07
5/30: obj: 5536.49
5/31: tv: 0.00
5/32: xfm: 106697.66
5/33: obj: 3189.01
5/34: tv: 0.00
5/35: xfm: 33437.75
5/36: obj: 3143.08
5/37: tv: 0.00
5/38: xfm: 25480.56
5/39: obj: 3140.55
5/40: tv: 0.00
5/41: xfm: 24579.75
5/42: obj: 3140.29
5/43: tv: 0.00
5/44: xfm: 24474.94
5/45: obj: 3140.26
5/46: tv: 0.00
5/47: xfm: 24462.69
5/48: obj: 3140.25
5/49: tv: 0.00
5/50: xfm: 24461.25
5/51: obj: 3140.25
5/52: tv: 0.00
5/53: xfm: 24461.09
5/54: obj: 3140.25
5/55: tv: 0.00
5/56: xfm: 24461.07
5/57: obj: 3140.25
5/58: tv: 0.00
5/59: xfm: 24461.07
5/60: obj: 3140.25
5/61: tv: 0.00
5/62: xfm: 24461.07
5/63: obj: 3140.25
5/64: tv: 0.00
5/65: xfm: 24461.07
 6/1: ls
 6/2: import transforms as tf
 6/3: from __future__ import division
 6/4: import numpy as np
 6/5: import scipy as sp
 6/6: import matplotlib.pyplot as plt
 6/7: import os.path
 6/8: import transforms as tf
 6/9: import scipy.ndimage.filters
6/10: import grads
6/11: import sampling as samp
6/12: import direction as d
6/13: from scipy import optimize as opt
6/14: EPS = np.finfo(float).eps
6/15: filename =
6/16: filename = '/home/asalerno/Documents/pyDirectionCompSense/data/SheppLogan256.npy'
6/17: filename
6/18: strtag = ['spatial','spatial']
6/19: TVWeight = 0.01
6/20: XFMWeight = 0.01
6/21: im = np.load(filename)
6/22: im
6/23: plt.imshow(im)
6/24: plt.show()
6/25: N
6/26:  N = np.array(im.shape)
6/27: N = np.array(im.shape)
6/28: %paste
6/29: pctg = 0.25
6/30: strtag
6/31: im_TV = tf.TV(im,N,strtag,dirWeight = 0)
6/32: im_TV.shape
6/33: grads.gTV(im)
6/34: grads.gXFM(im,N)
6/35: im.shape
6/36: im.shape = im.size
6/37: im.shape = 1,im.size
6/38: im.shape = (1,im.size)
6/39: im.shape
6/40: im.size
6/41: im.shape = 1,65536
6/42: im.flatten
6/43: im.shape
6/44: im.flatten()
6/45: im.shape
6/46: im.ravel()
6/47: im.shape
6/48: im = im.ravel()
6/49: im.shape
6/50: grads.gXFM(im,N)
6/51: im.shape
6/52: im.reshape(256,256)
6/53: im.reshape(65535)
6/54: im.reshape(65536)
6/55: im.;shape
6/56: im.shape
6/57: im.shape = 65536,1
6/58: im.shape
6/59: im.reshape(256,256)
6/60: im.shape
6/61: a = im.reshape(256,256)
6/62: a is im
6/63: im_TV
6/64: np.tanh(0+0j)
6/65: np.tanh(0+1j)
6/66: x
6/67: x = np.linspace(-5,5)
6/68: x
6/69: x = np.linspace(-5,5,1000)
6/70: y0 = np.tanh(x)
6/71: y1 = np.tanh(0.5*x)
6/72: y2 = np.tanh(2*x)
6/73: plt.plot(x,y0,x,y1,x,y2)
6/74: plt.show()
6/75: plt.plot(x,y0,x,y1,x,y2)
6/76: plt.legend(['1','0.5','2'])
6/77: plt.show()
6/78: np.tanh(im_TV)
6/79: plt.imshow(np.tanh(im_TV))
6/80: plt.imshow(abs(np.tanh(im_TV)))
6/81: im_TV.shape
6/82: plt.imshow(abs(np.tanh(im_TV)[i,:,;]))
6/83: plt.imshow(abs(np.tanh(im_TV)[i,:,:]))
6/84: plt.imshow(abs(np.tanh(im_TV)[1,:,:]))
6/85: plt.show()
6/86: im_TV
6/87: im_TV.shape
6/88: a = np.sum(im_TV,axis=0)
6/89: a.shape
6/90: import rwt.wavelets as wv
6/91: wv.WAVELETS_DICT
6/92: wv.WAVELETS_DICT()
6/93: wv.WAVELETS_DICT
6/94:
def dot_product_threshold_with_weights(filename,
                          threshold = 0.1,
                          sigma = 0.35/2):
        dirs = np.loadtxt(filename)
        num_vecs = dirs.shape[0]
        cnt = 0
6/95:     locs = np.array([])
6/96:     vals = []
6/97:
    for i in xrange(0,num_vecs):
            for j in xrange(1,num_vecs):
                    dp = np.dot(dirs[i,:],dirs[j,:])
6/98:
            if dp >= threshold:
                    cnt = cnt + 1
                    locs[cnt,:] = np.vstack([locs, [i, j]])
                    vals[cnt] = np.vstack([vals, np.exp((dp**2 - 1)/(2*sigma**2))])
6/99:     return locs, vals
6/100: %paste
6/101: %paste
6/102: dot_product_with_mins('/home/asalerno/Documents/pyDirectionCompSense/GradientVectorMag.txt')
6/103: dirs = load('GradientVectorMag.txt')
6/104: dirs = np.load('GradientVectorMag.txt')
6/105: dirs = loadtxt('GradientVectorMag.txt')
6/106: dirs = np.loadtxt('GradientVectorMag.txt')
6/107: dirs
6/108: inds = dot_product_with_mins(dirs)
6/109: inds
6/110: inds.shape
6/111: inds.index(3)

figuresCode.py
import os
from sys import path as syspath
os.chdir(
    '/home/asalerno/Documents/pyDirectionCompSense/')  # Change this to the directory that you're saving the work in
syspath.append("/home/asalerno/Documents/pyDirectionCompSense/source/")
import numpy as np
#import errorcalc as err
import matplotlib.pyplot as plt
import saveFig
import transforms as tf
plt.rcParams['image.cmap'] = 'gray'

def rmse(xk,x):
    return np.sqrt(np.sum((xk-x)**2)/len(xk))
    
def zpad(orig_data,res_sz):
    res_sz = np.array(res_sz)
    orig_sz = np.array(orig_data.shape)
    padval = (np.ceil((res_sz-orig_sz)/2))
    res = np.pad(orig_data,([int(padval[0]),int(padval[0])],[int(padval[1]),int(padval[1])]),mode='constant')
    return res

im = np.load('/home/asalerno/Documents/pyDirectionCompSense/brainData/exercise_irradiation/bruker_data/running_C/P14/fullySampledBrain.npy')
imgsData = np.zeros([5,6,6,4,294,294])
rmse_data = np.zeros([5,6,6,4])
pc=-1
TV=-1
XFM=-1
imdc=-1



for pctg in [0.20, 0.25, 0.33, 0.40, 0.50]:
    pc += 1
    for TVWeight in [0.0001, 0.0005, 0.001, 0.005, 0.01, 0.05]:
        TV+=1
        for XFMWeight in [0.0001, 0.0005, 0.001, 0.005, 0.01, 0.05]:
            XFM+=1
            for imdcs in ['zeros','ones','densCorr','imFull']:
                imdc += 1
                imgsData[pc,TV,XFM,imdc,:,:] = np.load("brainData/exercise_irradiation/bruker_data/running_C/P14/rad_0.1/TV%s_XFM%s/%.2fper_result_im_dc_" % (float('%.1g' % TVWeight), float('%.1g' % XFMWeight) , pctg) + imdcs + ".npy")
                rmse_data[pc,TV,XFM,imdc] = rmse(imgsData[pc,TV,XFM,imdc,:,:],im)
            imdc=-1
        XFM=-1
    TV=-1

pctg = [0.20, 0.25, 0.33, 0.40, 0.50]
TVWeight = [0.0001, 0.0005, 0.001, 0.005, 0.01, 0.05]
XFMWeight = [0.0001, 0.0005, 0.001, 0.005, 0.01, 0.05]
imdcs = ['zeros','ones','densCorr','imFull']

N = len(XFMWeight)

ind = np.arange(N)+.5
width =0.15

# TV and XFM Comparison
# imdcs = densCorr -- 2
# pctg = 0.33 -- 2
fig, ax = plt.subplots()
rects1 = ax.bar(ind - 3*width, rmse_data[2,0,:,2], width, color='r', alpha=.1)
rects2 = ax.bar(ind - 2*width, rmse_data[2,1,:,2], width, color='r', alpha=.3)
rects3 = ax.bar(ind - 1*width, rmse_data[2,2,:,2], width, color='m', alpha=.5)
rects4 = ax.bar(ind + 0*width, rmse_data[2,3,:,2], width, color='m', alpha=.7)
rects5 = ax.bar(ind + 1*width, rmse_data[2,4,:,2], width, color='b', alpha=.8)
rects6 = ax.bar(ind + 2*width, rmse_data[2,5,:,2], width, color='b', alpha=1)

ax.set_ylabel('RMSE')
ax.set_title('RMSE for Different TV and XFM Combinations')
ax.set_xticks(ind)
ax.set_xticklabels(('0.0001', '0.0005', '0.001', '0.005', '0.01', '0.05'))
ax.set_xlabel('XFMWeights')
ax.legend((rects1[0], rects2[0], rects3[0], rects4[0], rects5[0], rects6[0]),('0.0001', '0.0005', '0.001', '0.005', '0.01', '0.05'),title='TVWeights',loc='upper left'); #loc='center left', bbox_to_anchor=(1, 0.5))


## im_dc and pctg Comparison
## TV = 0.005 -- 3
## XFM = 0.005 -- 3
#N = len(imdcs)

#ind = np.arange(N)+.5
#width =0.15

#fig, ax = plt.subplots()
#rects1 = ax.bar(ind - 3*width, rmse_data[0,3,3,:], width, color='b', alpha=.2)
#rects2 = ax.bar(ind - 2*width, rmse_data[1,3,3,:], width, color='b', alpha=.4)
#rects3 = ax.bar(ind - 1*width, rmse_data[2,3,3,:], width, color='b', alpha=.6)
#rects4 = ax.bar(ind + 0*width, rmse_data[3,3,3,:], width, color='b', alpha=.8)
#rects5 = ax.bar(ind + 1*width, rmse_data[4,3,3,:], width, color='b', alpha=1)

#ax.set_ylabel('RMSE')
#ax.set_title('RMSE for Different Pctg and Starting Point Combinations')
#ax.set_xticks(ind)
#ax.set_xticklabels(('zeros','ones','densCorr','imFull'))
#ax.set_xlabel('Percent Sampled')
#ax.legend((rects1[0], rects2[0], rects3[0], rects4[0], rects5[0], rects6[0]),('20%', '25%', '33%', '40%', '50%'), title='im_dc',loc='upper right'); #loc='center left', bbox_to_anchor=(1, 0.5))


# im_dc and pctg Comparison
# TV = 0.005 -- 3
# XFM = 0.005 -- 3
rmse_lr = np.zeros(len(pctg))
for i in range(len(pctg)):
lrLoc = int(np.ceil((N[0]-np.ceil(N[0]/np.sqrt(1/pctg)))/2))
data_lr_us = np.fft.fftshift(tf.ifft2c(im,np.ones(im.shape)))[lrLoc:-lrLoc,lrLoc:-lrLoc]
data_lr_rs = zpad(data_lr_us,im.shape)
im_lr_rs = abs(tf.fft2c(np.fft.fftshift(data_lr_rs),np.ones(im.shape)))
rmse_lr[i] = err.rmse(abs(im_lr_rs),im)

N = len(pctg)

ind = np.arange(N)+.5
width =0.15

fig, ax = plt.subplots()
rects0 = ax.bar(ind - 4*width, rmse_lr, width, color='b', alpha=0.1)
rects1 = ax.bar(ind - 3*width, rmse_data[:,3,3,0], width, color='b', alpha=.25)
rects2 = ax.bar(ind - 2*width, rmse_data[:,3,3,1], width, color='b', alpha=.5)
rects3 = ax.bar(ind - 1*width, rmse_data[:,3,3,2], width, color='b', alpha=.75)
rects4 = ax.bar(ind + 0*width, rmse_data[:,3,3,3], width, color='b', alpha=1)

ax.set_ylabel('RMSE')
ax.set_title('RMSE for Different Pctg and Starting Point Combinations')
ax.set_xticks(ind)
ax.set_xticklabels(('20%', '25%', '33%', '40%', '50%'))
ax.set_xlabel('Percent Sampled')
ax.legend((rects1[0], rects2[0], rects3[0], rects4[0]),('zeros','ones','densCorr','imFull'), title='im_dc',loc='upper right'); #loc='center left', bbox_to_anchor=(1, 0.5))



# Subplots for each XFM and TV solution and each pctg
pc=-1
TV=-1
XFM=-1
N = im.shape
minval = np.min(im)
maxval = np.max(im)

for pctg in [0.20, 0.25, 0.33, 0.40, 0.50]:
    pc += 1
    lrLoc = int(np.ceil((N[0]-np.ceil(N[0]/np.sqrt(1/pctg)))/2))
    im_lr = tf.fft2c(np.fft.fftshift(np.fft.fftshift(tf.ifft2c(im,np.ones(im.shape)))[lrLoc:-lrLoc,lrLoc:-lrLoc]),np.ones(im[lrLoc:-lrLoc,lrLoc:-lrLoc].shape))
    for TVWeight in [0.0001, 0.0005, 0.001, 0.005, 0.01, 0.05]:
        TV+=1
        for XFMWeight in [0.0001, 0.0005, 0.001, 0.005, 0.01, 0.05]:
            XFM+=1
            fig = plt.figure(figsize=(15,10))
            ax1 = fig.add_subplot(2,3,1)
            plt.imshow(im, clim=(minval,maxval))
            ax1.set_title('Original Image')
            ax2 = fig.add_subplot(2,3,2)
            plt.imshow(imgsData[pc,TV,XFM,0,:,:],clim=[minval,maxval])
            ax2.set_title('Recon %.2f%% samp imdc=zeros' % (pctg*100))
            ax3 = fig.add_subplot(2,3,3)
            plt.imshow(imgsData[pc,TV,XFM,1,:,:],clim=(minval,maxval))
            ax3.set_title('Recon %.2f%% samp imdc=ones' % (pctg*100))
            ax4 = fig.add_subplot(2,3,4)
            plt.imshow(abs(im_lr))
            ax4.set_title('Low Res with %.2f%% of Data' % (pctg*100))
            ax5 = fig.add_subplot(2,3,5)
            plt.imshow(imgsData[pc,TV,XFM,2,:,:],clim=(minval,maxval))
            ax5.set_title('Recon %.2f%% samp imdc=densCorr' % (pctg*100))
            ax6 = fig.add_subplot(2,3,6)
            plt.imshow(imgsData[pc,TV,XFM,3,:,:],clim=(minval,maxval))
            ax6.set_title('Recon %.2f%% samp imdc=imFull' % (pctg*100))
            
            saveFig.save("brainData/exercise_irradiation/bruker_data/running_C/P14/rad_0.1/TV%s_XFM%s/%.2fper_im_lr_edit_comparison" % (float('%.1g' % TVWeight), float('%.1g' % XFMWeight), pctg))
        XFM=-1
    TV=-1

